server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -500,7 +500,11 @@ public String toString() {
         }
 
         long getFreeBytes() {
-            return diskUsage.getFreeBytes() - relocatingShardSize;
+            try {
+                return Math.subtractExact(diskUsage.getFreeBytes(), relocatingShardSize);
```

Comment:
How could this overflow?
Isn't `relocatingShardSize >= 0` and `0 <= diskUsage.getFreeBytes() <= Long.MAX_VALUE`? Is this to capture the case where `diskUsage.getFreeBytes() == 0 && relocatingShardSize == Long.MAX_VALUE`?

Reply:
No, `relocatingShardSize` is the net size of all relocations, i.e. shards in minus shards out, so can indeed be negative :(